File: 0116-populating-next-right-pointers-in-each-node/0116-populating-next-right-pointers-in-each-node.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node left;
    public Node right;
    public Node next;

    public Node() {}
    
    public Node(int _val) {
        val = _val;
    }

    public Node(int _val, Node _left, Node _right, Node _next) {
        val = _val;
        left = _left;
        right = _right;
        next = _next;
    }
};
*/

class Solution {
    // Time complexity O(n) and space O(n)
    public Node connect(Node root) {
        if(root == null)
            return null;
        Queue<Node> bfs = new LinkedList<>();
        bfs.add(root);
        while(!bfs.isEmpty()) {
            var size = bfs.size();
            for(int i = 0; i < size; i++) {
                var current = bfs.poll();
                if(i != size - 1) {
                    current.next = bfs.peek();
                }

                if(current.left != null) {
                    bfs.add(current.left);
                }

                if(current.right != null) {
                    bfs.add(current.right);
                }
            }
        }

        return root;
    }
}

File: 0116-populating-next-right-pointers-in-each-node/0116-populating-next-right-pointers-in-each-node.py
"""
# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next
"""
from collections import deque

## Time complexity O(n) and space complexity O(n)
class Solution:
    def connect(self, root: 'Optional[Node]') -> 'Optional[Node]':
        if not root:
            return None

        bfs = deque([root])

        while bfs:
            size = len(bfs)
            for i in range(size):
                current = bfs.popleft()
                if i != size - 1:
                    current.next = bfs[0]  # Peek at the next element in the queue

                if current.left:
                    bfs.append(current.left)

                if current.right:
                    bfs.append(current.right)

        return root